server/src/controllers/course.ts

```
@@ -40,3 +42,118 @@ export async function addCourse(req: Request, res: Response): Promise<void> {
+  Exam = 'EXAM'
+}
+
+function validateTeachingMethod(teachingMethod: any): teachingMethod is TeachingMethod {
+  return typeof teachingMethod === 'string' && (
+    teachingMethod === 'LECTURE' ||
+      teachingMethod === 'EXAM'
+  );
+}
+
+interface CourseInstanceAddRequest {
+  gradingType: GradingType;
+  startingPeriod: Period;
+  endingPeriod: Period;
+  teachingMethod: TeachingMethod;
+  responsibleTeacher: number;
+  startDate: string;
+  endDate: string;
+}
+
+function validateCourseInstanceAddFormat(body: any): body is CourseInstanceAddRequest {
+  return body &&
+    body.gradingType &&
+    body.startingPeriod &&
+    body.endingPeriod &&
+    body.teachingMethod &&
+    body.responsibleTeacher &&
+    body.startDate &&
+    body.endDate &&
+    validateGradingType(body.gradingType) &&
+    validatePeriod(body.startingPeriod) &&
+    validatePeriod(body.endingPeriod) &&
+    validateTeachingMethod(body.teachingMethod) &&
+    typeof body.responsibleTeacher === 'number' &&
+    typeof body.startDate === 'string' &&
+    typeof body.endDate === 'string';
+}
+
+export async function addCourseInstance(req: Request, res: Response): Promise<void> {
+  try {
+    const courseId: number = Number(req.params.courseId);
+
+    if (!validateCourseInstanceAddFormat(req.body)) {
+      throw new Error('Invalid course instance addition format');
+    }
+
+    const request: CourseInstanceAddRequest = req.body;
+
+    const course: Course | null = await models.Course.findOne({
+      where: {
+        id: courseId
+      },
+    });
+
+    if (course == null) {
+      throw new Error(`Course with ID ${courseId} does not exist`);
+    }
+
+    models.CourseInstance.create({
+      courseId: courseId,
+      gradingType: request.gradingType,
+      startingPeriod: request.startingPeriod,
+      endingPeriod: request.endingPeriod,
+      teachingMethod: request.teachingMethod,
+      responsibleTeacher: request.responsibleTeacher, // TODO: Is this fine?
```

Works as long as the responsible teacher exists in our users table. request with non existing id will throw error, for an example "'Key (responsible_teacher)=(-1) is not present in table "user".'". 

Maybe we could check if the teacher actually exists with something like: 
```javascript
const teacher: User | null = await models.User.findOne({ where: { responsibleTeacher: request.responsibleTeacher } });
if (teacher === null) {
  throw new Error(`Responsible teacher with ID ${request. responsibleTeacher} does not exist`);
} else {
  // create a new instance
}
```